### PR TITLE
exception handling & ext props

### PR DESCRIPTION
**Pull request by jeremeguenther**
- Replacing Enterprise library exception handler with native copy
- New Extended property sql feature allowing extended sql attributes to be used to decorate classes.
- regex sql injection bug fix
- adding data to thrown exceptions
- error handling default error message bug fix

### CODE REVIEW

File: Source/Entities/Validation/CommonRules.cst [hunk 1]
```diff
@@ -857,6 +888,73 @@ namespace <%=NameSpace%>.Validation
...
         #endregion		
+
+        /// <summary>
+        /// Assigns an error string, and calls a handler if it exists
+        /// </summary>
+        /// <param name="ruleType"></param>
+        /// <param name="target"></param>
+        /// <param name="e"></param>
+        /// <param name="err"></param>
+        private static void AssignError(CommonRuleType ruleType, object target, ValidationRuleArgs e, string err)
```
Comment (jeremeguenther) [inline Source/Entities/Validation/CommonRules.cst:899]: This method was created to provide a central place for event handling.  I am still running off the event handling I submitted for a pull request last time though, and am only about half way through writing a subscription based event handling system.  So for now I have removed the event code from this method.

**Review by rollercoaster — APPROVED** [review]
👌

File: Source/ASP.NET/Common/WebConfig.cst [hunk 2]
```diff
@@ -47,6 +47,7 @@ string entlibVersionText = GetEntLibVersionSignature(EntLibVersion);
 		enableEntityTracking="false"
         enableMethodAuthorization="false"
         defaultCommandTimeout="30"
+		exceptionHandlerType=""
```
Comment (niemyjski) [inline Source/ASP.NET/Common/WebConfig.cst:50]: I'm curious why this couldn't be inferred at generation time

File: Source/DataAccessLayer/Bases/NetTiersProviderBase.cst [hunk 3]
```diff
@@ -93,7 +100,36 @@ namespace <%=DALNameSpace%>.Bases
...
+
+                        try
+                        {
+                            System.Reflection.Assembly.Load(exceptionHandlerTypeStringAssembly);
+                            exceptionHandlerType = Type.GetType(exceptionHandlerTypeString, false, true);
+                        }
+                        catch
+                        {
+                            //throws file not found exception
+                        }
```
Comment (niemyjski) [inline Source/DataAccessLayer/Bases/NetTiersProviderBase.cst:124]: I never like just eating exceptions. Can we log it, or just handle the file not found exception?

File: Source/DataAccessLayer/Bases/NetTiersProviderBase.cst [hunk 4]
```diff
@@ -93,7 +100,36 @@ namespace <%=DALNameSpace%>.Bases
...
                 if (entityCreationalFactoryType == null)
                     throw new ArgumentNullException("Could not find a valid entity factory configured in assemblies.  .netTiers can not continue.");
-                
+
+                /*
+				 * Attempt to create the Exception Handler
+				 */
+                if (!string.IsNullOrEmpty(exceptionHandlerTypeString))
+                {
+                    //exceptionHandlerType = Type.GetType(exceptionHandlerTypeString, false, true);
```
Comment (niemyjski) [inline Source/DataAccessLayer/Bases/NetTiersProviderBase.cst:109]: dead code

File: Source/DataAccessLayer/ExceptionManager.cst [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            get
+            {
+                if (_handler == null && DataRepository.Provider.ExceptionHandlerType != null)
+                {
+                    lock (HandlerLock)
+                    {
+                        if (_handler == null)
+                        {
+                            if (typeof(IExceptionHandler).IsAssignableFrom(DataRepository.Provider.ExceptionHandlerType) == false)
+                                throw new System.Configuration.ConfigurationErrorsException("The Exception Handler type does not implement the IExceptionHandler interface");
```
Comment (niemyjski) [inline Source/DataAccessLayer/ExceptionManager.cst:32]: I think we should do this check before locking

File: Source/DataAccessLayer/Utility.cst [hunk 6]
```diff
@@ -307,7 +307,7 @@ namespace <%=DALNameSpace%>
 
         #region SqlInjection
         private static readonly System.Text.RegularExpressions.Regex regSystemThreats =
-                new System.Text.RegularExpressions.Regex(@"\s?;\s?|\s?drop\s|\s?grant\s|^'|\s?--|\s?union\s|\s?delete\s|\s?truncate\s|\s?sysobjects\s?|\s?xp_.*?|\s?syslogins\s?|\s?sysremote\s?|\s?sysusers\s?|\s?sysxlogins\s?|\s?sysdatabases\s?|\s?aspnet_.*?|\s?exec\s?",
+                new System.Text.RegularExpressions.Regex(@"\s?;\s?|\bdrop\b|\bgrant\b|^'|\s?--|\bunion\b|\bdelete\b|\btruncate\b|\bsysobjects\b|\s?xp_.*?|\bsyslogins\b|\bsysremote\b|\bsysusers\b|\bsysxlogins\b|\bsysdatabases\b|\baspnet_.*?|\bexec\b",
```
Comment (niemyjski) [inline Source/DataAccessLayer/Utility.cst:310]: What does `\b` do?

File: Source/DataAccessLayer/Utility.cst [hunk 7]
```diff
@@ -836,9 +844,8 @@ private void WriteExecuteMethod(ExecuteType executeType, bool useTransactions)
                     RetryMaxAttempts,
                     sleepType,
                     "DataSet",
-                    (useTransactions? ", transactionManager.TransactionObject":""),
-                    (useTransactions? "transactionManager.Database":"database"),
-                    (useTransactions? "if (!transactionManager.IsOpen) throw new DataException(\"Transaction must be open before executing a query.\");":""),
+                    string.Format(returnResults, "DataSet"),
+                    (useTransactions? "if (!transactionManager.IsOpen && !transactionManager.IsDisabled) throw new DataException(\"Transaction must be open before executing a query.\");":""),
```
Comment (niemyjski) [inline Source/DataAccessLayer/Utility.cst:848]: Can we break these into variables so it's easier to read what's being passed into the format.

File: Source/TemplateLib/CommonSqlCode.cs [hunk 8]
```diff
@@ -6575,6 +6575,20 @@ public enum CacheLayer : byte
...
+    #region Exception Handler
+    /// <summary>
+    /// Enterprise Library versions
+    /// </summary>
+    public enum ExceptionHandler : byte
+    {
+        /// <summary>Generate a basic error handling system that errors will be passed into</summary>
+        netTiers = 0,
+        /// <summary>Use Enterprise Library version</summary>
+        EntLib = 1,
```
Comment (niemyjski) [inline Source/TemplateLib/CommonSqlCode.cs:6587]: This probably needs to be flipped, EntLib has been the default for decades.

File: Source/VisualStudio/vsnet2005.project.cst [hunk 9]
```diff
@@ -1107,6 +1107,10 @@
 			  <SubType>Code</SubType>
 			</Compile>
 
+			<Compile Include="<%=DALSubFolder%>ExceptionManager.cs">
```
Comment (niemyjski) [inline Source/VisualStudio/vsnet2005.project.cst:1110]: Isn't this conditional based on if you are not using ent lib?

File: Source/DataAccessLayer/Utility.cst (hunk 6 above)
Comment (jeremeguenther) [reply Source/DataAccessLayer/Utility.cst:310]: word break.  It includes spaces but also handles things like returns.

File: Source/Entities/Validation/CommonRules.cst [hunk 10]
```diff
@@ -857,6 +888,73 @@ namespace <%=NameSpace%>.Validation
...
+        {
+            e.Description = err;
+        }
+
+        /// <summary>
+        /// The validation rule type
+        /// </summary>
+        public enum CommonRuleType : int
```
[the remark this answers is not in the document]
Comment (jeremeguenther) [reply Source/Entities/Validation/CommonRules.cst:907]: Probably not yet.  I use this as part of that exception handling system I built in.  The one where you suggested it be converted to a subscription system.  Still working on that conversion.

File: Source/VisualStudio/vsnet2005.project.cst (hunk 9 above)
Comment (jeremeguenther) [reply Source/VisualStudio/vsnet2005.project.cst:1110]: That would be the cleaner way of writing it :-D

**Comment by niemyjski** [comment]
Is this ready to be merged?

**Pull request closed without merging**